controllers/machine.go

```diff
@@ -327,6 +331,12 @@ func (mrc *machineReconcileContext) ensureHardwareUserData(hardware *tinkv1.Hard
 }
 
 func (mrc *machineReconcileContext) ensureHardware() (*tinkv1.Hardware, error) {
+        // Make the process of picking up the hardware and updating it's ownership atomic.
+        // This avoids a race that could potentially cause picking up the same hardware 
+        // for different machines.
+        lock.Lock()
+        defer lock.Unlock()
```

Comment:
Perhaps lock could be released right after `takeHardwareOwnership` to make time of the locking a bit shorter?

Reply:
Exactly what I had in mind and had tested with. I will have to try this again to recall exactly what the issue was, but the best way to fix this was to ensure atomicity of entire hardware update operation, including the userdata setting and patching via setStatus.